apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,8 +1395,221 @@ private void StartPush()
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    var backoff = s_ackBackoff.NextBackoff(retryInfo.CurrentBackoff ?? TimeSpan.Zero);
+                    // We should retry only for specified timeout.
+                    // If current UTC time is greater than first UTC time of failure + timeout, we should not retry.
+                    if (DateTime.UtcNow > retryInfo.FirstTimeOfFailureInUtc + TimeSpan.FromSeconds(timeoutInSeconds) - backoff)
+                    {
+                        // We are past the retry timeout.
+                        return;
+                    }
+
+                    // If we have ids to retry and we are not past the specified timeout, then retry.
+                    // Update backoff of retryableErrorIds in _retryableIds dictionary.
+                    UpdateAllRetryableErrorBackoff(retryableErrorIds, backoff);
+                    // Retry errors that are still temporary errors. They will be retried after backOff.
+                    Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
+                    // TODO: Check, if the below expression can be simplified or optimized.
+                    Add(delayTask, new NextAction(false, hasAcksOrNacks
+                    ? () => { ackActionToRetry(retryableErrorIds); StartPush(); }
+                    : () => { extendActionToRetry(extendIds.Where(j => retryableErrorIds.Contains(j.Id))); StartPush(); }));
+                }
+
+                // Updates the backoff of the retryableErrorIds in _retryableIds dictionary.
+                void UpdateAllRetryableErrorBackoff(IEnumerable<string> ids, TimeSpan? backoff)
+                {
+                    foreach (var id in ids)
+                    {
+                        if (_retryableIds.TryGetValue(id, out var info))
+                        {
+                            _retryableIds[id] = info.WithBackoff(backoff);
+                        }
+                    }
+                }
+
+                // Updates the _retryableIds dictionary. Ensures that the ids that should be retried exist in the dictionary.
+                // By design, only ids that are in _retryableIds dictionary will be retried with exponential backoff.
+                // Removes the ids that are either successful or become permanent failures as they should not be retried.
+                void UpdateRetryableIds(IEnumerable<string> allIds, AckError ackError)
+                {
+                    var temporaryFailureIds = ackError?.TemporaryFailureIds ?? Enumerable.Empty<string>();
+                    var permanentFailureIds = ackError?.PermanentFailureIds ?? Enumerable.Empty<string>();
+                    // Successful Ids = AllIds - (TemporaryFailures + Permanent failures)
+                    var successfulIds = allIds.Locked(() => allIds.Except(temporaryFailureIds.Union(permanentFailureIds)));
```

Comment:
I don't think you need `Locked` here - and if you do, there's probably a bigger issue.

Reply:
I suspect it's also more efficient to write `allIds.Except(temporaryFailureIds).Except(permanentFailureIds);`